### ISSUE TITLE
Thoughts on Commit hooks 2.0

### ISSUE THREAD

**Issue by miyconst** [Starcounter/Starcounter#2760]
After short play with the Commit hooks I have some thoughts.

1. The `Hook<T>.OnDelete` should also pass an object, but not its ID. Commit hook with an ID makes it useless. The object is deleted by that time and there is no way to know what other entity or file was associated with this object.
2. The `Hook<T>.OnXxxx` should be called for every object of type `T` or `X : T`. Lets say I have a class `Image`, in `OnDelete` action I remove the physical file. Another app which uses same db has a class `Photo : Image` with some custom behaviour. The `OnDelete` hook should also be invoked for `Photo` objects.
3. Non of the hooks fires for objects created/updated inside a hook action. Not sure it should work, differently, just noting a fact.
4. Application fails with an exception `The operation failed because of an invalid object access.` if I delete just inserted object inside an `OnInsert` hook. This is probably correct, just noting another fact.
5. The hook names are misleading a bit and there is no description available. If the call back executes after an action, then the names should be `OnDeleted`, `OnInserted`, `OnUpdated`, if before the action then it should be `OnBeforeDelete`, `OnBeforeInsert`, `OnBeforeUpdate`. I think there should be all of those hooks.
6. How to detach a hook? Native c# `event` supports operators `+=` and `-=`, why not use it?
7. What if I need a hook for a single object only? I would rather implement an `IHookable` interface in my db class and then subscribe to the object's events. Some code to explain what I mean:
```csharp
public interface IHookable {
    void OnBeforeDelete();
    void OnBeforeUpdate();
    void OnBeforeInsert();
}

public class Product : IHookable {
    public event EventHandler BeforeInsert;
    public event EventHandler BeforeUpdate;
    public event EventHandler BeforeDelete;

    public string Name;
    public DateTime CreateDate;
    public DateTime ChangeDate;

    public void OnBeforeDelete() {
        if (this.BeforeDelete != null) {
            this.BeforeDelete(this, EventArgs.Empty);
        }
    }

    public void OnBeforeUpdate() {
        this.CreateDate = DateTime.Now;

        if (this.BeforeUpdate != null) {
            this.BeforeUpdate(this, EventArgs.Empty);
        }
    }

    public void OnBeforeInsert() {
        this.CreateDate = this.ChangeDate = DateTime.Now;

        if (this.BeforeInsert != null) {
            this.BeforeInsert(this, EventArgs.Empty);
        }
    }
}

public class StarcounterInternal {
    public void CommitTransaction() {
        foreach (var entity in this.EntitiesToInsert) {
            IHookable hook = entity as IHookable;

            if (hook != null) {
                hook.OnBeforeInsert();
            }

            this.InsertObject(entity);
        }
    }
}

public class Program {
    static void Main() {
        Db.Transact(() => {
            var p = new Product() { 
                Name = "I am a special product!"
            };

            p.BeforeInsert += (s, a) => {
                (s as Product).Name = "This is a hook for only one object, not entire set of products";
            };
        });
    }
}
```

**Comment by per-samuelsson** [comment]
Hi @miyconst,

Thanks for the feedback!

I will consider each item eventually, and comment on them, and after that we can consider which we should turn into real work items.

**Comment by per-samuelsson** [comment]
Starting with the most simple ones to answer/comment:

> 3) Non of the hooks fires for objects created/updated inside a hook action. Not sure it should work, differently, just noting a fact.

This is by design, to avoid recursive (and possibly infinite) behaviour. Yes, we probably **can** make it work, if we introduce something like a visited-table to keep track of what objects have been visited, and instead raise an error once we detect a circular trigger, but the current implementation is more trivial.

**Comment by miyconst** [comment]
I agree on the point 3. EntityFramework ORM works in the same way.

**Comment by per-samuelsson** [comment]
> 1) The Hook<T>.OnDelete should also pass an object, but not its ID. Commit hook with an ID makes it useless. The object is deleted by that time and there is no way to know what other entity or file was associated with this object.

I wouldn't go as far as to say it makes it useless. A replication/mirroring mechanism could use it, for example, propagating the delete to some slave(s).

Also, changing this would require a rewrite of the kernel transaction commit, since by the time we invoke the commit hooks, the proxy object is just not usable any more (semantically, it's already deleted).

Furthermore, we have complementing IEntity.OnDelete for the scenario you request. In that case, you get the call while the object is still around (even being a virtual instance method), and you can complement the current transaction with stuff like cascading deletes and make them atomic too, since they will be carried out only if the transaction actually commit.

And finally, I guess you still could find objects referencing the object being deleted by means of SQL and query on the key, or does our SQL not support such behaviour? Like doing a select on some table filtering on a reference column by passing the ID, rather than the object reference?

**Comment by per-samuelsson** [comment]
> 2) The Hook<T>.OnXxxx should be called for every object of type T or X : T. Lets say I have a class Image, in OnDelete action I remove the physical file. Another app which uses same db has a class Photo : Image with some custom behaviour. The OnDelete hook should also be invoked for Photo objects.

We have referencing to this as inheritable hooks, and the current implementation is by design: we choose to have registration per class/table, rather than propagating it down to childs. There were several reasons why.

The main reason was that we want to keep the "low-level" implementation as **simple** as possible, and the described behaviour can be achieved on a higher, application level, even though it would require some work from developers. What I mean is, that since we support an overload of every ```Hook<T>.OnXXX``` that allows a runtime type to be passed, developers can register basically every type they which at runtime.

Some pseudo code to show what I mean:

```C#
void Main() {
  foreach (var t in GetLoadedDatabaseTypes()) {
    Hook<object>.OnInsert(t, o => {
      Console.Write("An object {0}/{1} was inserted", o.GetObjectId(), o.GetType().FullName);
    }
  }
}
```

Connecting back to your example, just replace ```object``` with ```Image```, and register ```Photo``` (and other images on that hook), something like:

```C#
void Main() {
  foreach (var t in GetLoadedDatabaseTypes()) {
    if (typeof(Image).IsAssignableFrom(t)) {
      Hook<Image>.OnInsert(t, i => {
        Console.Write(i.Resolution);
      }
    } 
  }
}
```

A second reason was that since hooks are quite expensive in terms of performance, we wouldn't want them to be active if not absolutly neccessary. Hence, per table/class registration by default minimize the risk of that.

Finally, in 1.x we supported a notion of "filterable hooks", that was basically a built-in high level API built on top of hooks, that supported inherited registration/invocation like you suggest, but we haven't ported anything such to the current version.

**Comment by per-samuelsson** [comment]
> 4) Application fails with an exception ```The operation failed because of an invalid object access```. if I delete just inserted object inside an OnInsert hook. This is probably correct, just noting another fact.

Hmm, this sounds like a bug to me, but I'm not sure to be honest. I have to consider it some. What if you try a simple transaction (not involving hooks) on the same class and do:

```C#
Db.Transact(() => {
  new Foo().Delete();
}
```

Just from the top of my head, I think both should work equally, but maybe the shouldn't...

**Comment by per-samuelsson** [comment]
> 6) How to detach a hook? [...]

There is currently no way to detach hooks, but we probaly should support it eventually. Right now, hooks are registered on a code host lifetime.

> 6) [..] Native c# event supports operators += and -=, why not use it?
> 7) The hook names are misleading a bit and there is no description available. If the call back executes after an action, then the names should be OnDeleted, OnInserted, OnUpdated, if before the action then it should be OnBeforeDelete, OnBeforeInsert, OnBeforeUpdate. I think there should be all of those hooks.

We could certainly consider refining the API's. And, after that, document them.

**Comment by miyconst** [comment]
>1) The Hook.OnDelete should also pass an object, but not its ID. Commit hook with an ID makes it useless. The object is deleted by that time and there is no way to know what other entity or file was associated with this object.

I need a hook which would be invoked just before deleting an object, so the current one could be renamed to `OnDeleted` and another one `OnBeforeDelete` should be introduced.

`IEntity.OnDelete` is available only in the assembly which defined the class. In our case this is `Simplified` project. I can't modify the method in each app I need the hook.
The only way to use it would be to add an event `BeforeDelete` to a class and fire the event inside the `OnDelete` method. Then in third-party app subscribe to the event, but this will work for a single object only, not for entire set.

And finally :), I can't select any related entities to the one being deleted, because the relation properties are emptied by that time.

Lets say in `Images` app I have history of files for an `Illustration`. 
I want to delete all of the history records and physical files when an `Illustration` deleted.
The class `Illustration` is defined in `Simplified` project, so the `OnDelete` method is not available for me.
`Hook<Illusration>.OnDelete` does not provide any information except the deleted object id. 
I do not have access to the physical file path, because the `Illustration` object has already been deleted.
And I also can't select history records, because the rows are no longer related to the `Illustration` and theirs foreign key property is `NULL`.

Correct me if I am wrong, but looks like with the current implementation the given example is not achievable.

**Comment by per-samuelsson** [comment]
Another comment regarding this:

> The hook names are misleading a bit and there is no description available. If the call back executes after an action, then the names should be `OnDeleted`, `OnInserted`, `OnUpdated`, if before the action then it should be `OnBeforeDelete`, `OnBeforeInsert`, `OnBeforeUpdate`. I think there should be all of those hooks.

Hooks are executed **as part of** a committing transaction, so I frankly dont know exactly what that means in the context of your comment.

Consider this pseudo code:

```C#
void Main() {
  Hook<Foo>.OnInsert(f => { throw new Exception(); });
  Db.Transact(() => {
    new Foo();
  });
}
```

In the above, **no new** `Foo` **will be inserted**. At the end of the transact scope, the hook will be invoked, and since the hook fail, the transaction will fail as a whole. The hook operates as part of it.

**Comment by miyconst** [comment]
>2) The Hook.OnXxxx should be called for every object of type T or X : T. Lets say I have a class Image, in OnDelete action I remove the physical file. Another app which uses same db has a class Photo : Image with some custom behaviour. The OnDelete hook should also be invoked for Photo objects.
The `foreach (var t in GetLoadedDatabaseTypes()) {` works for already loaded classes only. This would not have a practical use aas Polyjuice is a set of tons small apps, which can be started and stopped at any time.

In my example I assume that two apps are developed by two different developers. And the author of `Photo` class is not aware of any hooks for `Image` class.

An example: class `A` with properties `X`, `Y`, `Total { get { return X + Y; } }` and class `B : A { int Z; }`. This means that class `B` inherits *all* of the properties from class `A` and extends it with a property `Z`.
It would be ridiculous if property `Total` would stop working in class `B`. 

Same for classes `Image` & `Photo`. Developer inherits class `Photo` from `Image` and wants to receive *all* the functionality, but files for photos magically stay on disk even after photo is deleted.

If this is not technically limited, then we should let developer choose how the hook should behaviour. Fire for class and all subclasses or only for a particular class.

**Comment by miyconst** [comment]
@per-samuelsson should we start a new issue for each of the items? This one seems to grow too fast.

**Comment by miyconst** [comment]
This one works:

```csharp
Db.Transact(() => {
    new Foo().Delete();
}
```

But this one fails:

```csharp
Db.Transact(() => {
    var foo = new Foo();
    foo.Delete();
    foo.Name = "!!";
}
```

Lets say there are two `Hook<Something>.OnInsert` hooks. The first hook deletes the object, what the second one will receive? An exception will occur. It is logical for me. Entities can't be removed inside hooks.

**Comment by miyconst** [comment]
>In the above, **no new** `Foo` **will be inserted**. At the end of the transact scope, the hook will be invoked, and since the hook fail, the transaction will fail as a whole. The hook operates as part of it.

Okay. Then it seems like current hooks are already behaving like `OnBeforeUpdate` & `OnBeforeInsert`, but the `OnDelete` hook is very confusing. I can cancel the transaction by throwing an exception, but can't access the object, because it is already "deleted".

**Comment by per-samuelsson** [comment]
> [...] but the OnDelete hook is very confusing. I can cancel the transaction by throwing an exception, but can't access the object, because it is already "deleted".

Yes, I can see your point, and I agree to some level. At the same time, its consitent with how a proxy behave right after you have deleted it, like this:

```C#
var f = Db.SQL<Foo>("SELECT f FROM Foo f").First;
f.Delete();

// Will raise exception, since `f` has been deleted in the
// current context/transaction, hence you cant 
Console.Write(f.Bar);
```

**Comment by per-samuelsson** [comment]
> @per-samuelsson should we start a new issue for each of the items? This one seems to grow too fast.

Sounds like a good idea to me, but maybe we shouldn't create one for all 7 points, but rather try to see where they intersect and make some assesment on what issues are actually to be created.

I'm also working on #2762 in parallel to this now, and as part of that, I'm trying different ideas on how to make the API simpler to understand and use.
